File: app/lua/token_bucket.lua
local key = KEYS[1]

local capacity = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])

local now = redis.call("TIME")
local current = tonumber(now[1]) + tonumber(now[2]) / 1000000

local tokens = redis.call("HGET", key, "tokens")
local last_refill = redis.call("HGET", key, "last_refill")

if tokens == false or tokens == nil then
    redis.call(
        "HSET",
        key,
        "tokens",
        capacity - 1,
        "last_refill",
        current
    )

    redis.call(
        "EXPIRE",
        key,
        math.ceil(capacity / refill_rate) + 1
    )

    return 1
end

tokens = tonumber(tokens)
last_refill = tonumber(last_refill)

local elapsed = current - last_refill
local new_tokens = elapsed * refill_rate

tokens = math.min(
    capacity,
    tokens + new_tokens
)

last_refill = current

local allowed = 0

if tokens >= 1 then
    tokens = tokens - 1
    allowed = 1
end

redis.call(
    "HSET",
    key,
    "tokens",
    tokens,
    "last_refill",
    last_refill
)

redis.call(
    "EXPIRE",
    key,
    math.ceil(capacity / refill_rate) + 1
)

return allowed

File: aegis/algorithms/token_bucket.py
import time

from aegis.algorithms.base import RateLimiterAlgorithm
from aegis.core.token_bucket_state import TokenBucketState
from aegis.storage.base import Storage

class TokenBucket(RateLimiterAlgorithm):
    def __init__(
        self,
        capacity: int,
        refill_rate: float,
        storage: Storage[TokenBucketState]
    ):
        self.capacity = capacity
        self.refill_rate = refill_rate
        self.storage = storage
        
    def allow(self, key: str) -> bool:
            current_time = time.time()
            
            state = self.storage.get(key)
            
            if state is None:
                state = TokenBucketState(
                    tokens=float(self.capacity),
                    last_refill=current_time,
                )
    
            elapsed = current_time - state.last_refill
            state.tokens = min(
                self.capacity,
                state.tokens + elapsed * self.refill_rate,
            )
    
            state.last_refill = current_time
            allowed = False

            if state.tokens >= 1:
                state.tokens -= 1
                allowed = True

            self.storage.set(key, state)
            return allowed